File: Tutorial3-1Scope1/src/main/java/com/example/tutorial3scope1/ThirdActivity.java
package com.example.tutorial3scope1;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.example.tutorial3scope1.di.DaggerUserComponent;
import com.example.tutorial3scope1.di.UserComponent;
import com.example.tutorial3scope1.model.User;

import javax.inject.Inject;

public class ThirdActivity extends AppCompatActivity {

    @Inject
    User user;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Get Application Component
        // ((MyApplication) getApplication()).getUserComponent().inject(this);

        // Get Class Component
        UserComponent daggerUserComponent = DaggerUserComponent.create();
        // TODO NEED to ADD void inject(ThirdActivity thirdActivity); UserComponent
        daggerUserComponent.inject(this);

        Button button = findViewById(R.id.button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(ThirdActivity.this, FourthActivity.class);
                startActivity(intent);
            }
        });


        TextView textView = findViewById(R.id.textView);
        textView.setText("Third User: " + user.hashCode());
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Toast.makeText(this, "ThirdActivity onDestroy()", Toast.LENGTH_SHORT).show();
    }
}

File: Tutorial5-3Dagger-Android-MultipleActivitiesAndFragments/src/main/java/com/example/tutorial5_3dagger_android_multipleactivitiesandfragments/third/ThirdActivity.kt
package com.example.tutorial5_3dagger_android_multipleactivitiesandfragments.third

import android.content.SharedPreferences
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.example.tutorial5_3dagger_android_multipleactivitiesandfragments.di.ActivityContributorModule
import com.example.tutorial5_3dagger_android_multipleactivitiesandfragments.model.SingletonObject
import com.example.tutorial5_3dagger_android_mutipleactivitiesandfragments.R
import dagger.android.AndroidInjection
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import javax.inject.Inject

/**
 * Must be added to [ActivityContributorModule] to not have
 * ERROR java.lang.IllegalArgumentException: No injector factory bound for Class<com.example
 */
class ThirdActivity : AppCompatActivity(), HasAndroidInjector {

    @Inject
    lateinit var dispatchingAndroidInjector: DispatchingAndroidInjector<Any>

    // Injected from AppModule with @Singleton
    @Inject
    lateinit var sharedPreferences: SharedPreferences

    // 🔥 Injected via Constructor Injection with @Singleton scope
    @Inject
    lateinit var singletonObject: SingletonObject


    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_third)

        findViewById<TextView>(R.id.tvInfo).text =
                "ApplicationModule sharedPreferences: ${sharedPreferences.hashCode()}\n" +
                        "Constructor @Singleton singletonObject: ${singletonObject.hashCode()}"
    }

    override fun androidInjector(): AndroidInjector<Any> {
        return dispatchingAndroidInjector
    }
}